### ISSUE TITLE
[FEA] Replace use of `rmm::device_vector` with `rmm::device_uvector` where possible

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#7287]
**Is your feature request related to a problem? Please describe.**

`rmm::device_vector` is an alias for `thrust::device_vector` that uses the default memory resource for allocation. Being an alias, we are limited to the function signatures of whatever `thrust::device_vector` provides. The problem is that `thrust::device_vector` does not take stream arguments for operations that involve allocations/copies/writes. This leads to problems like https://github.com/rapidsai/cudf/issues/2631.

**Describe the solution you'd like**

Ideally we'd have a `rmm::device_vector` that takes stream arguments for every stream ordered operation. However, that would be a significant effort to design and implement such a class. 

We may be able to just get away with using `rmm::device_uvector` instead, as it is its own type with explicit stream arguments for all stream ordered operations. My suspicion is that we can replace (almost) all uses of `rmm::device_vector` with `rmm::device_uvector` in libcudf (perhaps with some additional convenience stuff added to `rmm::device_uvector`). 

This issue will track the work to do the conversion of `device_vector` to `device_uvector` wherever possible. Here is a current list of modules that contain `device_vector` today (30 March 2021, branch-0.19): 

 - [x] [benchmarks](https://github.com/rapidsai/cudf/pull/8208)
 - [x] [gather](https://github.com/rapidsai/cudf/pull/7758)
 - [x] [merge](https://github.com/rapidsai/cudf/pull/8043)
 - [x] [scatter](https://github.com/rapidsai/cudf/pull/8074)
 - [x] [valid_if](https://github.com/rapidsai/cudf/pull/7758)
 - [x] [strings_column_view](https://github.com/rapidsai/cudf/pull/7798)
 - [x] sample
 - [x] [repeat](https://github.com/rapidsai/cudf/pull/8042)
 - [x] [groupby](https://github.com/rapidsai/cudf/pull/8148)
 - [x] [hashing](https://github.com/rapidsai/cudf/pull/8074)
 - [x] [join](https://github.com/rapidsai/cudf/pull/8074)
 - [x] count_elements (lists)
 - [x] [partitioning](https://github.com/rapidsai/cudf/pull/8074)
 - [x] [quantiles](https://github.com/rapidsai/cudf/pull/8076)
 - [x] [grouped_rolling](https://github.com/rapidsai/cudf/pull/8106)
 - [x] search
 - [x] is_sorted
 - [x] [rank](https://github.com/rapidsai/cudf/pull/8029)
 - [x] sort
 - [x] encode
 - [x] [tests](https://github.com/rapidsai/cudf/pull/8205)
 - [x] [cuIO: CSV](https://github.com/rapidsai/cudf/pull/7805)
 - [x] [cuIO: JSON](https://github.com/rapidsai/cudf/pull/8151)
 - [x] [cuIO: ORC](https://github.com/rapidsai/cudf/pull/7614)
 - [x] [cuIO: Parquet](https://github.com/rapidsai/cudf/pull/8151)
 - [x] [cuIO: parsing_utils](https://github.com/rapidsai/cudf/pull/8151)

**Describe alternatives you've considered**
The alternative is [designing and implementing](https://github.com/rapidsai/rmm/issues/120) a full-blown, `rmm::device_vector` class, but that's a lot more work.

**Comment by vuule** [comment]
Volunteered to replace device_vector in column_buffer. Will do this after #7397 is merged.

**Comment by davidwendt** [comment]
Volunteering to replace the `device_vector` in the `cpp/src/strings/` and `cpp/src/text/` source files (total about 20 files)

**Comment by harrism** [comment]
Working on replacing device_vector and std::vector with uvector and spans in concatenate.

**Comment by karthikeyann** [comment]
Volunteering to replace the `device_vector` in `groupby`. Will do this after #7387 is merged.
Sort based groupby is already done in #7523

**Comment by harrism** [comment]
Added a checklist. I'm working on gather.

**Comment by harrism** [comment]
@vuule did you check off ORC? Can you link the PR that completes it?

**Comment by vuule** [comment]
> @vuule did you check off ORC? Can you link the PR that completes it?

It's https://github.com/rapidsai/cudf/pull/7614, I see it linked above (15 days ago).

**Comment by harrism** [comment]
The only remaining use of `device_vector` in libcudf is in `span.hpp` and its tests.

Closing.
